## Alignment methods handling multiple genomics data generated from different single cells of the same tissue

Compared to multi-omics data, it is experimentally much easier to obtain multiple modalities of data where each modality is obtained from similar but different cells of the same tissue. The task to harmonize these data is called alignment (Figure {@fig:1}). The body of literature applying machine learning and statistical methods to this task is rich, including manifold learning, neural-network based methods, and Bayesian methods, as summarized in Table @tbl:2 and depicted in Figure {@fig:3}. Note that some of the methods developed for batch-correct different scRNA-seq datasets, could in principle be repurposed for single-cell multiple omics alignment; we refer readers to previous benchmark studies [@doi:10.1186/s13059-019-1850-9].

![Illustration of some common approaches for alignment of multi-omics single-cell data: Bayesian methods, manifold alignment methods and neural network based models.](images/Fig_3.png){#fig:3 width="75%" height="75%"}

| Methodology Category | Method      | Algorithm                                                            | Data                                              | Reference |
|----------------------|-------------|----------------------------------------------------------------------|---------------------------------------------------|-----------|
| Manifold Alignment   | UNION - Com | Topological Alignment                                                | Transcriptomic, Epigenetic                        | [@doi:10.1093/bioinformatics/btaa443]      |
|                      | MATCHER     | Pseudotime Reconstruction and Manifold Alignment                     | Transcriptomic, Epigenetic                        | [@doi:10.1186/s13059-017-1269-0]      |
|                      | MMD-MA      | Manifold Alignment                                                   | Transcriptomic, Epigenetic (DNAme)                | [@doi:10.1101/644310]      |
|                      | SCOT        | Gromov-Wasserstein optimal transport                                 | Transcriptomic, Epigenetic (DNAme, accessibility) | [@doi:10.1101/2020.04.28.066787]      |
|                      | Pamona      | Partial Gromov-Wasserstein optimal transport                         | Transcriptomic, Epigenetic                        | [@doi:10.1093/bioinformatics/btab594]      |
| Neural Network       | MAGAN       | Generative Adversarial Network                                       | Transcriptomic, Proteomic                         | [@doi:10.48550/arXiv.1803.00385]      |
|                      | SCIM        | Adversarial autoencoder                                              | Transcriptomic, Proteomic (CyTOF)                 | [@doi:10.1093/bioinformatics/btaa843]      |
|                      | Multigrate  | Variational Autoencoder                                              | Transcriptomic, Proteomic                         | [@doi:10.1101/2022.03.16.484643]      |
| Bayesian             | clonealign  | Bayesian latent variable model                                       | RNA-seq, DNA                                      | [@doi:10.1186/s13059-019-1645-z]      |
|                      | MUSIC       | Topic models                                                         | RNA, CRISPR                                       | [@doi:10.1038/s41467-019-10216-x]      |
|   Other              | Seurat v3   | Canonical Correlation Analysis and Mutual Nearest Neighbors analysis | RNA-seq, ATAC-seq                                 | [@doi:10.1016/j.cell.2019.05.031]      |
|                      | bindSC      |  Canonical Correlation Analysis                                      | RNA-seq, ATAC-seq                                 | [@doi:10.1101/2020.12.11.422014]      |
|                      | MAESTRO     |                                                                      | RNA-seq, ATAC-seq                                 | [@doi:10.1186/s13059-020-02116-x]      |
|                      | LIGER       | Matrix factorization                                                 | RNA-seq, methylation                              | [@doi:10.1016/j.cell.2019.05.031; @doi:10.1016/j.cell.2019.05.006]   |

Table: Summary of the computational methods for aligning multiple omics data from different single cells. {#tbl:2}

### Bayesian Methods

**Clonealign** [@doi:10.1186/s13059-019-1645-z] integrates single-cell RNA and DNA sequencing data from heterogeneous populations by assigning cells measured by RNA-seq to clones derived from DNA-seq data. Clonealign is based on a Bayesian latent variable model, where a categorical variable is used to specify cell assignment. The model maps the copy number of a gene to its expression value by introducing a copy number dosage effect on the gene expression. The model is also flexible enough to allow for additional covariates such as batch effects or biological information that can be inferred from the gene expression (cell cycle, etc.). In addition to simulation studies that demonstrated robustness, Clonealign was also applied on real cancer datasets to discover novel clone-specific dysregulated biological pathways.

**MUSIC** [@doi:10.1038/s41467-019-10216-x] is an unsupervised topic modeling method for integrative analysis of single-cell RNA data and pooled CRISPR screening data [@doi:10.1214/07-AOAS114]. The model links the gene expression profile of the cells and specific biological function by delineating perturbation effects,, allowing for better understanding of perturbation functions in single cell CRISPR data. In the perturbation effect prioritizing step, MUSIC utilizes the output from the topic model and estimates individual gene perturbation effects on cell phenotypes. It takes three different schemes in modeling combined single-cell and CRISPR data: an overall perturbation effect which represents the gene perturbation effect, a topic model which specifies the function of perturbation effectsway, and with respect to relationships between different perturbation effects. MUSIC was applied to 14 real single-cell CRISPR screening datasets and accurately quantified and prioritized the individual gene perturbation effect on cell phenotypes, with tolerance for substantial noise.

### Manifold Alignment Methods

Manifold alignment methods aim to infer a lower-dimensional structure within multiple complex datasets (Figure {@fig:3}B). Once this is done, points can be matched across the datasets. This is a very broad class of algorithms, and we here review several representative ones based on distinct ideas, such as the use of pseudotime trajectories, Kernel methods and distance-based matching of cells. The distance-based matching (Figure {@fig:4}) is a general idea containing several different realizations, such as **UNION-Com** [@doi:10.1093/bioinformatics/btaa443], **SCOT** [@doi:10.1101/2020.04.28.066787] and **Pamona** [@doi:10.1093/bioinformatics/btab594], which are reviewed below, among other methods.

![Summary of the distance-based alignment algorithm: cells are represented by nodes in two different graph representations and matched in order to preserve a notion of the distance on the graph.](images/Fig_4.png){#fig:4 width="75%" height="75%"}

**MATCHER** [@doi:10.1186/s13059-017-1269-0] is the first manifold alignment technique to align different forms of single-cell data. Their approach builds on trajectory inference [@doi:10.1038/nbt.2859]. It constructs pseudotime trajectories corresponding to cellular processes for each omic first, and then aligns them between different -omics. Pseudotime trajectory models the corresponding cellular process as a Gaussian process and infers the latent variable corresponding to pseudotime. This results in a set of curves capturing the biological processes, one for each -omics layer. Such curves are then projected onto a reference line so that different cells can be matched across -omics. The model makes a strong assumption that there is only one common biological process to be modeled.

**MMD-MA** [@doi:10.1101/644310], or Maximum Mean Discrepancy - Manifold Alignment, is a completely unsupervised method. The alignment is performed by matching low-dimensional representations of different -omics, constructed through a kernel-based technique that minimizes the MMD (Maximum Mean Discrepancy) [@{https://dl.acm.org/doi/10.5555/2188385.2188410}] between the two datasets. Additionally, the representations are constructed by taking into account the distortion of the distances in the original data while keeping the transformation as simple as possible. The model was evaluated on data containing gene expression and methylation values from the same single cells; the known cell correspondence information was hidden and MMD-MA was able to successfully reconstruct this information.

**UNION-Com** [@doi:10.1093/bioinformatics/btaa443] performs unsupervised alignment of different -omics datasets by matching the structure of the datasets. The idea is that, if different -omics layers indeed correspond to similar samples of cells, then the distance matrices of any two -omics layers will become very similar after rearranging the cell indices. A matching matrix connecting points across datasets is learned by optimizing the similarity of distance matrices after cell permutation. This approach of matching is an extension of GUMA ("Generalized Unsupervised Manifold Alignment") [@{https://dl.acm.org/doi/10.5555/2969033.2969098}] with newly allowed soft matchings. Subsequently, this method performs a version of t-SNE [@{http://www.jmlr.org/papers/v9/vandermaaten08a.html}] adopted for multi-modal data represented in the same latent space. This approach takes the overall structure of all datasets into account while matching the cells, without the requirement of identical distributions of different modalities. UNION-Com compared favorably with Seurat v3 and MMD-MA when evaluated on the quality of labels transferred between gene expression, methylation and chromatin accessibility data.

**SCOT** [@doi:10.1101/2020.04.28.066787] is similar to UNION-Com in terms of distance comparison across the -omics layers. However, it is formulated as a different optimization problem per the theory of optimal transport. It starts by considering k-nearest neighbor graphs in different -omics layers and uses those to compute distances between cells from different -omics, like UNION-Com. The soft matchings are applied here as well, with points matched probabilistically across datasets. Unlike UNION-Com, such matchings are obtained by considering a version of optimal transport given by the Gromov-Wasserstein distance, which generalizes the "earth-mover" Wasserstein distance to optimal transport between different spaces [@doi:10.1007/s10208-011-9093-5]. SCOT compared favorably to MMD-MA and UNION-Com on several real and simulated datasets containing transcriptomic and epigenetic (DNAme or chromatin accessibility) data. The model contains only two hyperparameters, making it particularly simple to tune.

**Pamona** [@doi:10.1093/bioinformatics/btab594] uses a similar approach to SCOT, but with a modification of optimal transport based on Partial Gromov-Wasserstein distance [@doi:10.48550/arXiv.2002.08276], which accounts for data points that do not have appropriate matches across datasets. By doing so, the authors can allow for possible imperfect alignment between the datasets, tolerating cell types present in one dataset only. After the alignment is found, the data corresponding to different modalities is projected down to a dimensionally reduced space using Laplacian Eigenmaps [@doi:10.1162/089976603321780317]. Benchmarked on several datasets containing transcriptomic and epigenetic data, their model outperformed SCOT, MMD-MA and Seurat v3.

### Neural Network-Based Methods

Neural networks, including autoencoders and generative adversarial networks (GAN), have been used for the unsupervised task of the alignment of -omics datasets. Autoencoders have been described earlier. GANs typically consist of two parts: the generator network and the discriminator network. While the generator tries to produce outputs of a form resembling a certain target dataset, the discriminator learns the difference between the generator's outputs and the elements of the target dataset. In this section, we summarize the relevant neural network methods below.

**SCIM** [@doi:10.1093/bioinformatics/btaa843] builds on a multi-domain translation approach [@doi:10.48550/arXiv.1902.03515] to integrate multi-omics data in an unsupervised fashion. It uses a separate variational autoencoder for each modality in order to map the data onto reduced latent space representations. Such representations are then aligned to have a similar structure, by using a discriminator network in addition to autoencoders which learns to distinguish between the latent space representations of different -omics. The two autoencoders and the discriminator network are trained simultaneously, resulting in the two latent spaces being maximally alike. Once both datasets are encoded into approximately corresponding representations, the points with similar latent representations are matched across the datasets. This model was tested on simulations from PROSSTT ("Probabilistic Simulation of Single-Cell RNA-seq Tree-Like Topologies") [@doi:10.1093/bioinformatics/btz078] as well as datasets containing gene expression and proteins, and performed favorably to MATCHER when applied to simulated data exhibiting a complex cellular differentiation process.

**MULTIGRATE** [@doi:10.1101/2022.03.16.484643] uses a multi-modal variational autoencoder structure to project multi-omics data onto
a shared latent space. While somewhat similar to the scMVAE model [@doi:10.1093/bib/bbaa287], this framework brings additional flexibility and can be used for integration of the paired and unpaired single-cell data. Furthermore, this model can integrate data from a multi-omics assay such as CITE-seq with data from a single-omics assay such as scRNA-seq. Data corresponding to different -omics are first passed through separate neural networks, before being combined by the Product of Experts technique [@doi:10.1162/089976602760128018] to form the latent distribution. The decoder networks then aim to reconstruct all of the -omics from this unified representation. To better align cells, Maximum Mean Discrepancy is added to the loss function, penalizing the misalignment between the point clouds belonging to different assays. Their model was used for the creation of multi-modal atlases, and mapping a COVID-19 single-cell dataset onto a multi-modal reference.

**MAGAN** [@doi:10.48550/arXiv.1803.00385] utilizes generative adversarial networks (GANs) to align data from different domains. MAGAN uses two tied GANs to translate between the -omics layers, while tying their parameters and requiring that their combination maps any point onto itself. Namely, if the first generator maps data point A to data point B, then the second generator should map B back to A. It is conceptually very similar to the CycleGAN [@doi:10.48550/arXiv.1703.10593] model from computer vision, but with a key innovation that allowed it to more efficiently align and integrate single-cell data. The novelty here was noting that while the CycleGAN framework was very good at aligning the datasets in aggregate, it would not necessarily correctly match individual points. This is a particularly important problem for single-cell data. To address this problem, MAGAN is augmented with a correspondence loss measuring the difference between points before and after being mapped by generators. This model was tested on a variety of datasets, ranging from a simulated dataset to MNIST handwritten digits to molecular data. The method was applied to combine transcriptomic and proteomic data in single cells. The model was shown to meaningfully align the datasets even when the correspondence information was not available.

### Other Methods

**CCA** (Canonical Correlation Analysis) based methods reduce the dimensionality of data by selecting for the degrees of freedom that are correlated between the datasets. Seurat v3 [@doi:10.1016/j.cell.2019.05.031] combines CCA with network concepts in order to align and integrate single-cell multi-omics data. After performing the CCA, the algorithm identifies anchors between the datasets and scores the quality of those anchors. Anchors are identified by MNNs (mutual nearest neighbors), and their quality is scored by considering the overlap between the neighborhoods of anchors. Similar to Seurat v3, MAESTRO [@doi:10.1186/s13059-020-02116-x] also utilized canonical correlation analysis for the integration of transcriptomic and epigenetic data, and provided a comprehensive analysis pipeline. bindSC [@doi:10.1101/2020.12.11.422014] also uses canonical correlation analysis to construct shared representations of the data, iteratively optimized using a custom procedure.

**LIGER** [@doi:10.1016/j.cell.2019.05.006] performs an iNMF (integrative non-negative matrix factorization) to learn factors explaining the variation within and across datasets. Data such as DNA methylation are first aggregated over genes. Cells corresponding to different datasets are described by separate sets of cell-specific factors. Gene factors consist of two components: one that is shared across datasets and one that is dataset specific; the model aims to make the dataset-specific portion as small as possible. After performing the matrix factorization, the shared factor neighborhood graph is formed, in which cells are connected based on the similarity of their factors, and used for aligning the cells across modalities. Recently, this nonnegative matrix factorization approach has been extended to incorporate the idea of online learning. It iteratively updates the model in real-time, and leads to better scalability and computational efficiency [@doi:10.1038/s41587-021-00867-x].
